File: src/main/java/com/freya02/botcommands/internal/prefixed/TextCommandComparator.java
package com.freya02.botcommands.internal.prefixed;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.prefixed.BaseCommandEvent;
import com.freya02.botcommands.api.prefixed.CommandEvent;
import com.freya02.botcommands.internal.utils.Utils;
import org.slf4j.Logger;

import java.lang.reflect.Method;
import java.util.Comparator;
import java.util.List;

public class TextCommandComparator implements Comparator<TextCommandInfo> {
	private static final Logger LOGGER = Logging.getLogger();

	@Override
	public int compare(TextCommandInfo o1, TextCommandInfo o2) {
		final Method o1CommandMethod = o1.getMethod();
		final Method o2CommandMethod = o2.getMethod();

		if (o1CommandMethod == o2CommandMethod) return 0;

		if (o1CommandMethod.getParameterTypes()[0] == BaseCommandEvent.class
				&& o2CommandMethod.getParameterTypes()[0] == CommandEvent.class) {
			return -1;
		}

		final int order1 = o1.getOrder();
		final int order2 = o2.getOrder();
		if (order1 != 0 && order2 != 0) {
			if (order1 == order2) {
				LOGGER.warn("Method {} and {} have the same order ({})",
						Utils.formatMethodShort(o1.getMethod()),
						Utils.formatMethodShort(o2.getMethod()),
						order1);
			}

			return Integer.compare(order1, order2);
		}

		final List<? extends TextCommandParameter> o1Parameters = o1.getOptionParameters();
		final List<? extends TextCommandParameter> o2Parameters = o2.getOptionParameters();

		for (int i = 0; i < Math.min(o1Parameters.size(), o2Parameters.size()); i++) {
			if (o1Parameters.get(i).getBoxedType() == o2Parameters.get(i).getBoxedType()) {
				continue;
			}

			if (o1Parameters.get(i).getBoxedType() == String.class) {
				return 1;
			} else {
				return -1;
			}
		}

		return 1;
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/prefixed/TextCommandComparator.kt
package com.freya02.botcommands.internal.commands.prefixed

import com.freya02.botcommands.api.parameters.RegexParameterResolver
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.commands.prefixed.autobuilder.metadata.TextFunctionMetadata
import com.freya02.botcommands.internal.parameters.ResolverContainer
import com.freya02.botcommands.internal.utils.ReflectionUtils.nonInstanceParameters
import com.freya02.botcommands.internal.utils.ReflectionUtils.shortSignatureNoSrc
import mu.KotlinLogging
import kotlin.math.min
import kotlin.reflect.KParameter
import kotlin.reflect.jvm.jvmErasure

internal class TextCommandComparator(private val context: BContextImpl) : Comparator<TextFunctionMetadata> {
    private val logger = KotlinLogging.logger {  }

    private val TextFunctionMetadata.optionParameters
        get() = func.nonInstanceParameters
            .drop(1)
            .filter { context.getService<ResolverContainer>().getResolverOrNull(it) is RegexParameterResolver<*, *> }

    //TODO is this correct ? the same old implementation did not function with the new objects.
    override fun compare(o1: TextFunctionMetadata, o2: TextFunctionMetadata): Int {
        if (o1.func == o2.func) return 0

        //Put command with options first
        if (o1.optionParameters.any() && !o2.optionParameters.any()) {
            return -1
        } else if (!o1.optionParameters.any() && o2.optionParameters.any()) {
            return 1
        }

        val order1 = o1.annotation.order
        val order2 = o2.annotation.order
        if (order1 != 0 && order2 != 0) {
            if (order1 == order2) {
                logger.warn(
                    "Method {} and {} have the same order ({})",
                    o1.func.shortSignatureNoSrc,
                    o2.func.shortSignatureNoSrc,
                    order1
                )
            }

            return order1.compareTo(order2)
        }

        val o1Parameters: List<KParameter> = o1.optionParameters
        val o2Parameters: List<KParameter> = o2.optionParameters
        for (i in 0 until min(o1Parameters.size, o2Parameters.size)) {
            if (o1Parameters[i].type == o2Parameters[i].type) {
                continue
            }

            return when (o1Parameters[i].type.jvmErasure) {
                String::class -> 1
                else -> -1
            }
        }

        return 1
    }
}